unstable/wlr-data-control-unstable-v1.xml

```diff
@@ -68,9 +69,18 @@
         appropriate destroy request has been called.
       </description>
     </request>
+
+    <!-- Version 2 additions -->
+
+    <event name="primary_selection" since="2">
+      <description summary="advertise primary selection support">
+        This event can be sent when binding to the wlr_data_control_manager
+        global to advertise that it supports the primary selection.
+      </description>
+    </event>
```

Comment:
This was probably discussed and I'm probably dumb for bringing it up

But isn't the existence of the primary selection global enough of an advertisement?

Reply:
It was discussed somewhat (see above). In brief, the problems with this is that there's no _the_ primary selection global; there are two protocols that implement primary selection as of now and there may be more in the future. The compositor may (only) support e.g. `zwp_primvary_selection_v5` whereas a tool that intends to use this only knows about `gtk_primary_selection`. The tools would have to resort to hacks like `strstr("primary")` which is relatively ugly compared to wlr-data-control stating what it supports explicitly.

Reply:
I consider gtk_primary_selection deprecated and all clients/compositors should be moving towards zwp. Future improvements to primary selection should land in zwp as well. We don't need to consider the needs of gtk_primary_selection here at all, I think.

Reply:
It's not about `gtk_primary_selection` itself, it's about what happens if the client and the compositor don't have a match in the (version of) the primary selection protocol they support. They could still both support wlr-data-control and do primary selection stuff through that, but there'd be no (sane) way for them to know.

And this is not a theoretical possibility; wl-clipboard doesn't support or know about `wp_primary_selection` if it's built with `wayland-protocols` v1.16 or earlier (which is what current Linux distros ship); if you remove `gtk_primary_selection` from Sway we're going to get into this very situation.

Reply:
I think the solution in that case is to send a patch updating whichever end is out of date. @emersion thoughts?

Reply:
More context here: https://github.com/swaywm/wlr-protocols/pull/34#issuecomment-458548155

This event makes error reporting friendlier, but at the same time the compositor could've decided not to set the selection for whatever reason. I'm fine with both tbh.

Reply:
My vote is in favor of removing this.

Reply:
Supposing this ends up being removed, what exactly would you recommend a client wishing to know whether this protocol is suitable for managing primary selection to do?

For example, you could say that one can manage primary selection using this protocol _if and only if_ there is a global named `wp_primary_selection_unstable_v1` (or something). In which case it'd be better if this fact was mentioned in the protocol.

Reply:
I don't see another reliable way of figuring out whether the compositor supports primary selection. Besides, why would we _want_ to rely on hacks when this event (which falls in scope of the protocol) solves the issue?

Reply:
Look for the global. This isn't a hack.

Reply:
_Which_ global is "the" global? `wp_primary_selection_unstable_v1`? `gtk_primary_selection` _or_ `wp_primary_selection_unstable_v1`? _Any_ global whose name includes "primary"?

I still thinks it's a bad idea, but if you decide to do it this way, you'll have to document it in the protocol, perhaps like this:

```
This request can only be made if a global with name wp_primary_selection_unstable_v1 is available. Otherwise, a protocol error will be raised.
```

or

```
This request will only have any effect if a global with name wp_primary_selection_unstable_v1 is available. Otherwise, it will be silently ignored.
```

Reply:
We already went over this. wp-primary-selection, whichever is the latest version.

Reply:
And we can just silently ignore it if the client makes the request, we needn't document that.

Reply:
Just to be clear, let me restate the points.

Utilities like `wl-copy` and `wl-paste` and other clipboard managing tools want to know whether the primary selection is supported to show an error message in case it isn't, instead of silently setting or retrieving an empty selection (which would be confusing to the user).

The `primary_selection` event in `data-control` is one way of solving this issue.

Your suggestion is that the presence of a common primary selection protocol global should be used instead. At the current time that would be `zwp_primary_selection_device_manager_v1`.

However, we argue that it isn't optimal, because when `wp-primary-selection` gets a breaking change or stabilizes, the name of the global will change. This would break any clipboard manager's primary selection support until the new name is added into its checks, a new release is drafted, is picked up by distributions, etc. Essentially it means that clipboard utilities relying on `data-control` will break and need a new release due to an update of a protocol that they _don't even use_.

Reply:
Yes, because if the unstable version changes, it's a different protocol entirely. Once it becomes stable the global name will not change. This is by design.

Reply:
The issue is that clipboard managers would only care about the _presence_ of the `wp-primary-selection` main global as an indicator for primary selection support (_not_ for operation, clipboard managers use `data-control` for all operation). Breaking changes of `wp-primary-selection` in no way affect the clipboard managers, yet under your proposal they would still require an update and a new release if `wp-primary-selection` has a breaking change or stabilizes.

Reply:
I've said my piece. Maybe @emersion can make the call.

Reply:
Hmm. In the end I'd prefer removing this. The long-term solution will be to check for `wp-primary-selection`'s global, so this event won't be needed. In the meantime you can add a check for `zwp-primary-selection-v*`. Also, if there is any major issue with this approach it's always easier to add a new event than to remove it.